### DISCUSSION TITLE
Proposal: switch on System.Type

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#356] [General]
@alrz commented on [Tue Jan 26 2016](https://github.com/dotnet/roslyn/issues/8194)

When it comes to runtime type comparison, [there are multiple ways to do so](http://blogs.msdn.com/b/vancem/archive/2006/10/01/779503.aspx), it's more cumbersome if you want to do it multiple times with `if else`. it would be really nice if we could literally `switch` on `System.Type` or `System.RuntimeTypeHandle` and compiler choose the fastest way, like using `TypeHandle`.

``` cs
Type type = ...;
switch(type) {
  case typeof(Foo): ... break;
  case typeof(Bar): ... break;
}
```

---

@GeirGrusom commented on [Tue Jan 26 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-175426114)

It's planned that you can [switch on type](https://github.com/dotnet/roslyn/blob/future/docs/features/patterns.md) in C# 7.

I think it will look like this:

``` csharp
switch(baz)
{
    case Foo foo:
        foo.DoFoo();
        break;
    case Bar bar:
        bar.DoBar();
        break;
}
```

---

@alrz commented on [Wed Jan 27 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-175546805)

@GeirGrusom That's not what I'm talking about.

---

@orthoxerox commented on [Wed Jan 27 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-175687563)

Well, it would be nice if types were first-class citizens, but that would require CLR support, I think.

---

@alrz commented on [Wed Jan 27 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-175688708)

@orthoxerox I even suspect that this behavior can be neatly implemented as an extension `is` operator (if that would be possible). So, no CLR support is required at all.

---

@HaloFour commented on [Wed Jan 27 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-175716309)

The CLR is fully capable of handling the logic.  The question is how such a statement should be expanded.  I'd imagine it would produce the same IL as a series of `if`/`else` statements since the IL `switch` opcode only works when the values are a range starting at 0 which the runtime type handle would not be.

``` cs
Type type = obj.GetType();
switch (type)
{
    case typeof(Foo):
        Console.WriteLine("Foo!");
        break;
    case typeof(Bar):
        Console.WriteLine("Foo!");
        break;
    case typeof(Baz):
    default:
        Console.WriteLine("Baz or Something Else!");
        break;
    case typeof(Frob):
        Console.WriteLine("Frob!");
        break;
}
```

effectively equivalent to:

``` cs
Type type = obj.GetType();
if (typeof(Foo).IsEquivalentTo(type))
{
    Console.WriteLine("Foo!");
}
else if (typeof(Bar).IsEquivalentTo(type))
{
    Console.WriteLine("Bar!");
}
else if (typeof(Frob).IsEquivalentTo(type))
{
    Console.WriteLine("Frob!");
}
else
{
    Console.WriteLine("Baz or Something Else!");
}
```

---

@alrz commented on [Wed Jan 27 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-175718455)

@HaloFour `switch` with pattern-matching already turns to mix of `if`, `else` and `switch` (at least this is what F# does). So I assume this can be done in a similar way. Also, I don't know about `IsEquivalentTo` but comparing `TypeHandle` causes no type load and will be emitted directly to IL (see the link in the opening post).

---

@HaloFour commented on [Wed Jan 27 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-175727090)

@alrz In IL it's all conditional branch opcodes, unless the condition meets the incredibly narrow requirements to make it work with `switch`.  I assume that ADTs in F# have an incrementing tag field which do meet those requirements.

If `if (typeof(Frob).TypeHandle.Equals(type))` is appreciably more efficient then sure, that's probably what the compiler should emit.

---

@drewnoakes commented on [Thu Jan 28 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-176100436)

Should this only match exact types? What about derived types? Would `case typeof(object)` catch everything or only `new object()`? Either way it's going to surprise someone somewhere.

---

@alrz commented on [Thu Jan 28 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-176101952)

@drewnoakes That would be actually nice but then it will depend on the order and we might lose `TypeHandle` optimization.

---

@AdamSpeight2008 commented on [Thu Jan 28 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-176123567)

Or it'll turn it into a dictionary lookup, similar to VB.net's `Select Case`

---

@CnSimonChan commented on [Sat Apr 09 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-207867184)

Seems you can do it in future branch
![snip_20160410064405](https://cloud.githubusercontent.com/assets/1330321/14406886/acd14546-fee7-11e5-9784-e38da4ba3f61.png)
But the compiler will crash when compiling it :(

---

@alrz commented on [Sat Apr 09 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-207932835)

@CnSimonChan I think that's because `switch` is already relaxed to accept any expression type (depending on the branch you're on) and this doesn't seem to be something that you'd get it "for free" via pattern-matching. So I think they should make it explicitly an error or consider implementing it.

---

@gafter commented on [Mon Apr 11 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-208447659)

@CnSimonChan Reported as #10459

---

@gafter commented on [Mon Apr 11 2016](https://github.com/dotnet/roslyn/issues/8194#issuecomment-208455851)

Since we have to fix the compiler crash, lets just make this work.

---

@weitzhandler commented on [Mon Mar 06 2017](https://github.com/dotnet/roslyn/issues/8194#issuecomment-284390496)

A compelling feature I come a lot across with. I'd love to see this implemented.

**Comment by Spongman** [comment]
I just ran into this and was really surprised that, given c#7's new switch statement stuff, that this simple case was still not covered.

there _is_ a rather long-winded workaround, however:
````c#
Type type = ...;
switch(type) {
  case Type t when t == typeof(Foo): ... break;
  case Type t when t == typeof(Bar): ... break;
}
````

**Comment by YairHalberstadt** [comment]
@gafter

Do you think this could be championed? You indicated above that you thought it was worth doing, and I think it fits nicely in with pattern matching.

I would be happy to implement it.

**Reply by davidnmbond** [reply]
@gafter  @YairHalberstadt please let me know if I can help!

**Comment by gafter** [comment]
@YairHalberstadt I am not interested in championing this.

**Comment by YairHalberstadt** [comment]
@gafter

Understood. That's a pity. It's only a minor feature, but it comes up quite regularly. For example you suggested I use it in https://github.com/dotnet/roslyn/pull/32843#discussion_r264916537

EDIT:

I was talking nonsense. You didn't suggest that. You suggested something else that didn't compile for a different reason.

**Comment by gafter** [comment]
> For example you suggested I use it

That was not what I meant by "switching on the type".  I meant something like

``` c#
switch (someObject)
{
    case SomeType1 s1:
        ....
    case SomeType1 s2:
        ...
    case SomeType3 s3:
        ...
}
```

I'm working on our options for making this compile reliably to a constant-time dispatch in future versions (likely C# 9 or so) without any source changes.

**Comment by Happypig375** [comment]
My TypeConverters will look so much better when this is implemented!

**Comment by MgSam** [comment]
I do this more operation far more than I do any other scenario where a switch statement/expression would be useful. And I always have to use `if` statements. I hope the LDT triages this issue at some point.

**Comment by soenneker** [comment]
Following - also interested in this feature.

**Reply by pinkfloydx33** [reply]
Same. This came up in an SO post I commented on the other day. Would be nice to see, I can think of a few places I could clean up if this were implemented.

**Reply by ManuelKugelmann** [reply]
I quite regularly stumble upon this missing feature - it would be a great addition to c#!

**Reply by BreyerW** [reply]
While i personally dont but when i do i really wish for cleaner switching on types too. Giant if elses with multiple typeof looks awful and after certain threshold becomes hard to read

**Comment by anaximander23** [comment]
Just hit this doing some stuff with custom model binding in ASP.NET; I have a value which is `string`, and I have a `ModelMetadata` instance with a property `UnderlyingOrModelType` that is an instance of `Type`. If I could just do `switch (prop.UnderlyingOrModelType)` with cases for each type that needs special handling, it'd be great.

**Comment by mburbea** [comment]
I've come across scenarios where this can be used as an acceptable replacement if you know the type in question is a struct. e.g.
```cs
public string M<T>() where T:struct => default(T) switch {
   byte => "byte",
   sbyte => "sbyte"
   _ => "Not a byte or sbyte"
};
```
Unfortunately, doesn't really work when all you have is a type and you want to switch as ref types default is null.

**Reply by Tinister** [reply]
Try `public string M<T>() => System.Runtime.Serialization.FormatterServices.GetUninitializedObject(typeof(T)) switch  { ...` for all types.  😉

**Reply by mburbea** [reply]
well the advantage of the above approach is that is that it get optimized away into nothing.
https://sharplab.io/#gist:e62ff8b898ce2f31fc13783bde391278

When you use FormatterService it generates a lot more code::
https://sharplab.io/#gist:e905ea465977d5c88ede6e7ed4443c43

**Comment by wolgemoth** [comment]
This is a feature I have been waiting for as I use this exact pattern in my current project. I prefer switch statements to else if chains because they very clearly communicate their purpose (whereas an else if chain can be ambiguous at times). 

Seeing how they both compile into basically the same thing, I can't imagine this functionality being too hard to add in.

**Comment by MgSam** [comment]
FYI- I just noticed Github Discussions doesn't seem to have a way to sort by "Thumbs Up" of the original post - you have to separately up vote the issue itself up. So to keep visibility on this issue - I suggest that all the people asking for it do that in addition to the thumbs up.

![image](https://user-images.githubusercontent.com/1965570/110512204-00836d00-80d3-11eb-9d6e-8c7c1c880cc8.png)

**Comment by bernd5** [comment]
You can do it now but with a slightly different syntax:

```CSharp
using System;

Type type = ...;
switch(type) {
  case var x when x == typeof(int):
     break;
  case var x when x == typeof(string):
     break;
}
```

In my eyes this is sufficent.

And it works for switch expressions, too:
```CSharp
var typename = GetValue()?.GetType() switch
{
    var type when type == typeof(int) => "int",
    var type when type == typeof(string) => "string",
    _ => throw new NotSupportedException()
};
```

In CSharp 8 we introduced the "Constant-Pattern". Perhaps you could propose a "Non-Constant-Pattern" with a link to this discussion.

`System.Type` is a very well known immutable type - but it could be nice for other immutable types like records, too.
But in general one should avoid the usage of runtime-type-objects if it is doable with compile-time-types.

**Reply by BreyerW** [reply]
Hardly sufficient. All i can say good thing about it is, it works. But it is actually more verbose than traditional giant if/else especially when one needs to stack conditions on top of these (very rare yes but not unlikely) so 0 improvement for this particular case.

**Reply by jnm2** [reply]
The constant pattern was introduced in C# 8, by the way.

**Comment by msedi** [comment]
Does the suggestion also take into account code elision? Currently in generic methods checks against a typeof() is completely elided if possible. The switch with the type should also take this into account.

**Reply by GeirGrusom** [reply]
> 
> 
> Does the suggestion also take into account code elision? Currently in generic methods checks against a typeof() is completely elided if possible. The switch with the type should also take this into account.

Isn't that a JIT issue rather than a compiler issue?

**Reply by bernd5** [reply]
Which suggestion? If a type check is statically know (at generic function instantiation) code can be eliminated by the compiler / JITer.
But here `type` is a runtime type - how should the jitter know that such checks are always false?

**Reply by msedi** [reply]
@bernd5: I was assuming the switch should work on typeof() which is currently supported by the JIT to be elided if the statement is done wit if..then.else checks. I think it should work the same way, the switch should degenerate to one single case (the switch is indeed removed), if this is different to if..then..else no one will recognize the difference. 

But I'm not sure if I fully understood the proposal.

**Reply by bernd5** [reply]
@msedi
The possibilities to elide some cases / code in case of runtime-type checks are very limited.
You could elide for example: typeof(T) == typeof(int) when T is known.
But this seems to be nearly impossible with arbitrary `System.Type`'s - they are more or less just a pointer, like any other object.

**Reply by msedi** [reply]
@bernd5. Yes. You are right. I didn't see this.

**Comment by davidnmbond** [comment]
This issue was linked to from here: https://www.reddit.com/r/csharp/comments/q5zsbr/switcht

The requested example:

```
private bool TypeIsCool<T>()
    => T switch
    {
        string => true,
        bool => false,
        _ => throw new NotSupportedException($"{typeof(T).Name} is not supported")
    };
```

This would be great syntactic sugar for the following best answer in that thread:

```
static bool TypeIsCool<T>()
    => typeof(T) switch
    {
        _ when typeof(T).Equals(typeof(string)) => true,
        _ when typeof(T).Equals(typeof(bool))=> false,
        _ => throw new NotSupportedException($"{typeof(T).Name} is not supported")
    };
```

This doesn't look to difficult to implement, the meaning is clear and it is narrow in scope (just the switch statement).

**Reply by DavidArno** [reply]
> This doesn't look to difficult to implement, the meaning is clear and it is narrow in scope (just the switch statement).

The problem with that is that it's more of a hack than a true solution as it's not then a true pattern. For example, I have a genuine line in my code that I'd love to express as:
```c#
if (T is T1 or T2 or T3 or T4) ...
```
which this proposal wouldn't cover. In addition, I don't want to have to write your solution in the way you do as the `throw` isn't making use of pattern matching. I'd want to express it as:
```c#
private bool TypeIsCool<T>()
    => T switch
    {
        string => true,
        bool => false,
        var t => throw new NotSupportedException($"{t.Name} is not supported")
    };
```
The best solution I've come up with so far to work around this missing feature is to introduce a new `Type<T>`:
```c#
public readonly struct Type<T>
{
    private static readonly Type<T> Instance = new();
    private static readonly Type TType = typeof(T);
    public static Type<T> Value { get; } = Instance;

    public string Name => TType.Name;
}
```
This then allows your example to be written as:
```c#
static bool TypeIsCool<T>()  => Type<T>.Value switch {
    Type<string> => true,
    Type<bool> => false,
    var t => throw new NotSupportedException($"{t.Name} is not supported")
};
```
This approach let's me express that `if` I mentioned at the start as:
```c#
if (Type<T>.Value is Type<T1> or Type<T2> or Type<T3> or Type<T4>) ...
```
It's not perfect as too much boxing of the struct occurs for my liking and it's still too "noisy", but it possibly points the way to a true pattern-based solution for testing `Type` values and generic parameters.

**Comment by Khaos66** [comment]
I was facing this use case today as well and created #5687 but was directed to this discussion.
```c#
private int ToInt() => _value switch {
   int i => i,
   long l => (int)l,
   _ => default
}

private int ToLong() => _value switch {
   int i => i,
   long l => l,
   _ => default
}

public T As<T>()
{
   T x = default;
   return x switch
   {
      int when ToInt() is T i => i,
      long when ToLong() is T l => l,
     _ => default
   }
}
```

But I'm more concerned about performance. `T x = default` is a killer, when `T` is a complex struct, that can't be `null`.

Thinking outside the box
---
Maybe instead of having one generic method with a switch, we could have multiple methods?
```c#
public int As<T>() where T is int 
{ ... }

public long As<T>() where T is long 
{ ... }
```

**Reply by bernd5** [reply]
What you could do today (without allocating a large struct on the stack) is:

```csharp
using System;

#nullable enable

IMyConvertibel c = new Container<int>(12);
var l = c.As<long>();
Console.WriteLine(l);
var i = c.As<int>();
Console.WriteLine(i);

interface IMyConvertibel{
	T? As<T>();
}

class Container<X> : IMyConvertibel
{
    X _value;

    public Container(X value)
    {
        _value = value;
    }

    private int ToInt() => _value switch
    {
        int i => i,
        long l => (int)l,
        _ => default
    };


    private long ToLong() => _value switch
    {
        int i => i,
        long l => l,
        _ => default
    };

    public T? As<T>()
    {
        return typeof(T) switch
        {
            var t when t == typeof(int) => Cast(ToInt()),
            var t when t == typeof(long) => Cast(ToLong()),
            _ => throw new NotImplementedException($"As<{(typeof(T).Name)}")
        };
		
        T Cast<S>(in S s){
	        ref var refVar = ref System.Runtime.CompilerServices.Unsafe.AsRef<S>(s);
	        ref var refT = ref System.Runtime.CompilerServices.Unsafe.As<S, T>(ref refVar);
	        return refT;
        }
    }
}
```

**Reply by quetzalcoatl** [reply]
public int As<T>() where T is int 
public long As<T>() where T is long 

Wouldn't it effectivelly pull half of C++ template specialization into C#?
The other half being SFINAE, which we also don't have in C# at the moment.
That all would be awesome, but IMHO won't going to ever happen in C# :)

**Comment by ThaDaVos** [comment]
So it seems the above hasn't got any solution still - I ran into this when working on creating a custom DataContractResolver - sadly `typeof` cannot be used inside the switch expression and the other patterns aren't that useful especially as I am trying to swtich as follows:
```C#
    public override bool TryResolveType(Type type, Type? declaredType, DataContractResolver knownTypeResolver, out XmlDictionaryString? typeName, out XmlDictionaryString? typeNamespace)
    {
        return (type, declaredType) switch
        {
            (DateOnly, DateOnly) => resolveTypeToXML(typeof(DateOnly), out typeName, out typeNamespace),
            _ => knownTypeResolver.TryResolveType(type, declaredType, knownTypeResolver, out typeName, out typeNamespace)
        };
    }
```

Gives me the error `An expression of type 'Type' cannot be handled by a pattern of type 'DateOnly'.`

**Reply by DavidArno** [reply]
Currently, pattern matching in C# is limited to comparing variables to types and constants. `typeof(T)` isn't a constant, so there's currently no solution to it other than a when guard.

There's hope though. Work around unions throws up challenges over how to pattern match them. An accidental side effect of this could be solutions to pattern matching `Type`, generic parameters and generally active patterns. Time will tell there.

**Reply by hughesjs** [reply]
@DavidArno - But *why* can't we make `typeof(T)` constant in cases where `T` is clearly known at compile time?

**Reply by hughesjs** [reply]
@ThaDaVos - The workaround I've been using is to switch on typeof(T).name and use `nameof(T)` as the case which is compile time constant.

```
        var res = typeof(T).Name switch
            {
                nameof(TypeA) => //etc
            };

```

It's not perfect, but as a workaround it gets the job done well enough.

**Reply by DavidArno** [reply]
@hughesjs, that is indeed a really nice solution. It's not perfect as you say as the types in the switch branches have to be known at compile time, but I suspect cases where they are not are edge cases, so it covers most scenarios.

**Reply by DavidArno** [reply]
> @DavidArno - But _why_ can't we make `typeof(T)` constant in cases where `T` is clearly known at compile time?

That is indeed a very good question. Not being a member of the language team, it's not one I can answer though. Sorry.

**Reply by Logerfo** [reply]
@hughesjs as there are always a lot of costs involved in the process of developing a compiler feature, the question should not be "why can't we make?", but "why should we make?" instead. Besides "why", we should also have a clear vision of "how" (since there are virtually infinite ways of achieving certain behavior and edge cases to be ruled) and "when" (since there are other feature requests to be considered).

**Reply by HermanEldering** [reply]
> @DavidArno - But *why* can't we make `typeof(T)` constant in cases where `T` is clearly known at compile time?

Maybe typeof(T) is not really constant/known at compile time?

The type depends on the assembly, and the assembly depends on what is loaded at runtime.

For instance you might compile referencing an assembly version x.y.1 but someone could replace that later by version x.y.2. In the new assembly the type could even have different members than previously.

The switch should only check some aspects of the type, probably similar to how variable and argument types are defined. Maybe that is namespace + type name?

The workaround with nameof might not be ideal because it would also allow types with the same name in another namespace.

Edit: it might be possible to create a workaround with code generators.

**Reply by msedi** [reply]
Maybe `FullName` would work better. Event better would be `AssemblyQualifiedName`. But I'm not sure this would help here, when using `nameof()`.

By the way, we once had a discussion aboiut performance and code elision here: https://github.com/dotnet/runtime/discussions/81150, maybe this is also of interest.

**Reply by HermanEldering** [reply]
> Maybe `Fullname` would work better. Event better would be `AssemblyQualifiedName`. But I'm not sure this would help here, when using `nameof()`.

My point was that `AssemblyQualifiedName` can't be constant since it depends on what is loaded at runtime.

Also to me it would be very confusing if someone would use that in a switch. Ie this could fail:

```
String foo = "foo";

switch (RuntimeAssemblyQualifiedNameOf(foo)) {
    CompileTimeAssemblyQualifiedNameOf(string): DoSomething();
    default: throw new Exception();
}
```

DoSomething() would become unreachable if the code runs on a different version of .net than it was compiled with.

**Reply by GeirGrusom** [reply]
It looks from the specification that the reason why type lacks support isn't whether it's considered a constant or not (after all, it _is_ supported in attributes, unlike decimal which you _can_ switch on). Switch just has a list of expected expression types, and `Type` [isn't one of them](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#1383-the-switch-statement) (though decimal isn't either).

To my simple mind it doesn't look like there's a meaningful difference between switching on `string` and `typeof` expression in this regard (`ldstr` vs `ldtoken`). It just simply isn't supported by the switch-statement.

**Reply by HaloFour** [reply]
For all of the technicalities of "what is a constant", the only thing preventing `switch` on `Type` from being implemented is enough interest in doing so.

**Reply by msedi** [reply]
@HaloFour: 

> is enough interest in doing so.

I agree, but when looking in the codebase (especially the ones that deal with math) are full of these statements that would for sure make the code more maintainable/readable. That's why I do not understand that the runtime/language team has not so much interest in improving this.

**Reply by HaloFour** [reply]
Interest would come from demonstrating value through use cases.  I wonder how many of those math use cases would be better solved through numeric generic constraints now, though.  I don't believe it would require any runtime support, it's more or less just syntax candy.

**Reply by msedi** [reply]
@HaloFour: 

> I wonder how many of those math use cases would be better solved through numeric generic constraints now

True for the math cases, but I think for example the `Vector<T>` implementation still require it since the individual intrinsics (Sse2, etc) are not generic itself.

**Reply by DavidArno** [reply]
> For all of the technicalities of "what is a constant", the only thing preventing `switch` on `Type` from being implemented is enough interest in doing so.

This is true. Currently there's no real interest amongst the language team to address this. So we all sit and wait for the day when another team in MS suddenly has a need for it as it'll then likely happen soon after... 😇

**Reply by hughesjs** [reply]
> @hughesjs as there are always a lot of costs involved in the process of developing a compiler feature, the question should not be "why can't we make?", but "why should we make?" instead. Besides "why", we should also have a clear vision of "how" (since there are virtually infinite ways of achieving certain behavior and edge cases to be ruled) and "when" (since there are other feature requests to be considered).

Sorry, to clarify, my "why" was more if there were any reasons as to why it was theoretically impossible/impractical without huge language design changes that I was unaware of, not from a project feasibility standpoint.

**Reply by GabeSchaffer** [reply]
I think that the `switch` use case isn't as big as the general ability to use `typeof` expressions in patterns. I have a lot of expression processing code that has to do different things based on the types involved, and being able to match types in patterns would make my code a lot nicer.

Right now I have some code like this:
```C#
// look for bool Enumerable.Contains<T>(this IEnumerable<T> source, T value)
else if (expr is MethodCallExpression call)
{
    if (call.Method.Name == "Contains")
    {
        if (call.Method.DeclaringType == typeof(Enumerable))
        {
            var needle = call.Arguments[0];
            var haystack = call.Arguments[1];
```
but it would much prefer to write this:
```C#
// look for bool Enumerable.Contains<T>(this IEnumerable<T> source, T value)
else if (expr is MethodCallExpression {
    Method: { Name: "Contains", DeclaringType: typeof(Enumerable) },
    Arguments: [ var haystack, var needle ] })
```

Every time I write something like
```C#
node.Method is { Name: "op_Equality" or "op_Inequality" } method && method.DeclaringType == typeof(string)
```
I wish I could write it as
```C#
node.Method is { Name: "op_Equality" or "op_Inequality", DeclaringType: typeof(string) }
```

**Comment by Xyncgas** [comment]
Maybe we should first turn switch statement to F# switch statement, then in that prototype add support for switch to do type checking like we can in F#, and release them together under a new brand called the new switch

**Reply by ShalokShalom** [reply]
That sounds complicated to say the least?

**Comment by TahirAhmadov** [comment]
I also need this feature. One key aspect to remember here is that most of the code is already done for strings, so allowing types is literally just lifting a restriction. (Granted, adding typeof() to cases may be extra work. )

**Comment by RenderMichael** [comment]
This issue and many adjacent ones could be solved if `typeof()` was known to return a `const`-like result

**Reply by Xyncgas** [reply]
Like in C++
The problem is if you give a function an object at run time you have to do run time checking to know its type : who knows what you gave it then

**Comment by msedi** [comment]
Just out of curiosity, `Type` has a `GUID` property. Would it make sense to switch on this one?
I'm not sure how stable this is because I made the experience that in ReflectionOnlyLoads it makes a difference (!?).

**Reply by huoyaoyuan** [reply]
You can't switch on `Guid` either because it's not a constant. It's only available for COM interop types. Achieving the GUID also has heavy overhead.

**Comment by SetTrend** [comment]
Like to ref to similar discussion: [Shouldn't the result of the c# typeof() operator be considered "const"?](https://github.com/dotnet/roslyn/discussions/77464)

I would need this for handling classes equally that are similar (i.e., provide the same property names/types) but do not share the same base class hierarchy.

**Reply by TahirAhmadov** [reply]
What about `typeof(List<T>)` where `T` is a generic parameter type?

**Reply by SetTrend** [reply]
There are cases when a type is, so-called, "open" and other cases when a type is closed.

When a type is closed, I believe typeof(T) can be safely regarded being a constant.

**Reply by HaloFour** [reply]
> I would need this for handling classes equally that are similar (i.e., provide the same property names/types) but do not share the same base class hierarchy.

Why does that necessitate switching against an expression of `Type` vs. against the type of the instance itself?

**Reply by GabeSchaffer** [reply]
I don't know this poster's use case, but it's often the case that there is no instance to look at. For example, I have a lot of code that manipulates `Expression` instances as part of a LINQ provider. Every time I write code that looks at the `.Type` of an `Expression`, I wish I could `switch` on it or match it against a pattern. Of course, processing LINQ queries also often involves matching method signatures (like "is this the `Contains` method on `string`? or `List<T>`?") and many other places where I have to match types instead of instances.

My thought was always that if they can make an exception to the `const` rule to allow `Type` instances in attribute parameters, why not make similar exceptions in `switch` and patterns?
